sklearn/decomposition/_mf.pyx

```diff
@@ -0,0 +1,286 @@
+                         np.ndarray[double, ndim=1] bias_samples,
+                         np.ndarray[double, ndim=1] bias_features,
+                         np.npy_intp n_samples,
+                         np.npy_intp n_features,
+                         np.npy_intp n_components,
+                         int n_iter,
+                         int fit_intercept,
+                         double regularization,
+                         double learning_rate,
+                         object random_state,
+                         int verbose,
+                         int adagrad):
+
+    cdef np.npy_intp i, j, k
+    cdef double value, estimation, error
+
+    if adagrad:
+        feature_grad_norms = np.zeros(n_features)
+        sample_grad_norms = np.zeros(n_samples)
+
+        feature_bias_grad_norms = np.zeros(n_features)
+        sample_bias_grad_norms = np.zeros(n_samples)
+
+    indices = np.arange(data.shape[0])
+
+    for current_iter in range(n_iter):
+        random_state.shuffle(indices)
+
+        for value_index in indices:
+            value = data[value_index]
+            i = row[value_index]
+            j = col[value_index]
+
+            estimation = bias_samples[i] + bias_features[j] + np.dot(L[i, :], R[:, j])
+            error = estimation - value
+
+            L_grad = error * R[:, j] + regularization * L[i, :]
+            R_grad = error *  L[i, :] + regularization * R[:, j]
+
+            r_discount = 1
+            l_discount = 1
+
+            if adagrad:
+                feature_grad_norms[j] += np.dot(R_grad, R_grad)
+                sample_grad_norms[i] += np.dot(L_grad, L_grad)
+
+                r_discount = sqrt(feature_grad_norms[j]) + 1e-6
+                l_discount = sqrt(sample_grad_norms[i]) + 1e-6
+
+            R[:, j] -= learning_rate / r_discount * R_grad
+            L[i, :] -= learning_rate / l_discount * L_grad
+
+            if fit_intercept:
+                s_discount = 1
+                f_discount = 1
+
+                if adagrad:
+                    feature_bias_grad_norms[j] += error ** 2
+                    sample_bias_grad_norms[i] += error ** 2
+
+                    s_discount = sqrt(sample_bias_grad_norms[i]) + 1e-6
+                    f_discount = sqrt(feature_bias_grad_norms[j]) + 1e-6
+
+                bias_samples[i] -= learning_rate / s_discount * error
+                bias_features[j] -= learning_rate / f_discount * error
+
+
+        if verbose > 0:
+            _log_step(L, R, bias_features, bias_samples,
+                      row, col, data, n_iter, current_iter,
+                      regularization)
+
+
+cdef rr1(double* bias,
+         np.ndarray[double, ndim=1] w,
+         np.ndarray[double, ndim=1] b,
+         np.ndarray[double, ndim=2] X,
+         np.ndarray[double, ndim=1] y,
+         double lam, int n_iter,
+         int fit_intercept):
+
+    cdef int n_samples = X.shape[0]
+    cdef int n_components = X.shape[1]
```

Comment:
I figured out, by reading the paper, that `rr` is ridge regression and `rr1` the contribution from the ALS-1 paper. Is `rr1` just a coordinate descent ridge solver?  I guess this works for the ALS-1 paper because 1) they don't run it to convergence and 2) they use warm restarts which probably makes the errors go down over outer iterations.  But we would need to check the efficiency of these tricks with convergence plots and benchmarks, as the cited paper has a small number of citations (46 since 2010).